### ISSUE TITLE
Reproducible error calling LogFFE 

### ISSUE THREAD

**Issue by fingolfin** [gap-packages/recog#1]
While using recog, many times I have encountered errors like this:

```
gap> x in r;
Error, LogFFE: <z> must be a nonzero finite field element in
 log := LogFFE( det, z ); called from 
slpforelement( ri )( ri, x ) called from
SLPforElement( rifac, gg ) called from
slpforelement( ri )( ri, x ) called from
SLPforElement( rifac, gg ) called from
slpforelement( ri )( ri, x ) called from
...  at line 1 of *stdin*
you can replace <z> via 'return <z>;'
```

With some effort, I managed to find a test case with which I can always reproduce this error, see https://gist.github.com/fingolfin/5cd619b22026b08089ce

See also [issue-1.txt](https://github.com/gap-packages/recog/files/1548794/issue-1.txt)

**Comment by fingolfin** [comment]
Note that the matrix `x` at the end is a zero matrix. As @hulpke says: the matrix x is actually *not* contained in the group. But some code in recog is clearly written without this in mind, and thus lacks some important sanity checks.

Specifically, in this case, a block triangular form is detected, and homomorphisms are defined which use this form. So suppose I have the group consisting of all these matrices over GF(2):
```
/ * * 0 0 \
| * * 0 0 |
| * * * 0 |
\ * * * * /
```
Now I let recog loose on that group. Then I ask recog whether the following matrix `x` is contained:
```
/ 0 0 0 1\
| 0 0 1 0 |
| 0 1 0 0 |
\ 1 0 0 0/
```
Of course it it now, it's obvious. But what happens it that recog happily applies the group homomorphisms it defined. The first one (after an initial base change) amounts to "cut out the upper left 2x2 block and proejct to that". Which if of course perfectly valid for matrices in our group, but fails for the matrix `x`, as the projection is an all-zero matrix.

Indeed, this idea leads to this much smaller example triggering the bug:
```gap
gens := []; e:=Z(2)^0;
mat:=IdentityMat(4,GF(2)); mat[1][2]:=e; Add(gens, mat);
mat:=IdentityMat(4,GF(2)); mat[2][1]:=e; Add(gens, mat);
mat:=IdentityMat(4,GF(2)); mat[3][2]:=e; Add(gens, mat);
mat:=IdentityMat(4,GF(2)); mat[4][3]:=e; Add(gens, mat);
G:=Group(gens);
ri:=RecognizeGroup(G);
bad:=Reversed(IdentityMat(4,GF(2)));
bad in ri; # ERROR
```

In any case: Fixing this should not have to wait for verification support. Rather, the projection homomorphisms here should detect that the input matrix is not in its domain of definition (simply because the output matrix is not invertible), and return a failure.

**Comment by fingolfin** [comment]
I have a fix ready, I think.

**Comment by hulpke** [comment]
> the projection homomorphisms here should detect that the input matrix is not in its domain 

In this example the homomorphism can test this easily, but in other cases (say the reduction is to rewrite over a smaller field) this will be harder. We need to be careful that a potentially expensive test is only applied in situations where one really needs to test membership.

**Comment by fingolfin** [comment]
Well, the naive check I initially implemented I computing the determinant. Of course that's too slow for big examples, but it was the easiest way to do a quick proof of concept.

The proper implementation I want to write later today will verify that the block structure is satisfied. But this still requires scanning over the whole matrix (Note: if we ever start supporting sparse matrices, we'd want to do this differently, so perhaps this test should be moved into a separate operation `VerifyMatrixHasGivenBlockStructure` which could have better methods for sparse matrices).

Of course we could even bypass that test in situations were we know all inputs will definietly satisfy the requirement.

However, I am not so sure whether this (a) really is necessary (TODO: benchmark the effect this has on real-world recognition), and (b) whether it is feasible.

Here is why I bring up feasibility: Of course for a given set of generators, the meataxe gives us a definite answer for the block structure (modulo bugs in our implementation). So at the top level of our tree, that is fine, and also in the factors. But when you look at the kernel, then in many cases its generators are obtained by a randomized algorithm, and thus they oftern are initially to small. We detect that long before any verification would take place (I think). And during this detection, it could happen that we used an initial, incomplete set of kernel generators. If we then find another kernel element outside the subgroup generated by this incomplete generating set, then we would run into this kind of error again.

Hmmm.... I guess we could avoid that by tracking whether we are looking only at elements defined in terms of the current generating sets; or at elements from "external" sources, and only enable the verification in this case...

As to how to inform the homomorphisms whether to verify their input or not, we could for example use `PushOptions`. 

But I still would first like to test the actual performance impact of this. I'd like to avoid premature optimizations :).

**Comment by hulpke** [comment]
I don't disagree. My remark was simply meant to warn that we cannot rely only on homomorphism tests to keep us home dry.

**Comment by fingolfin** [comment]
Sure, fully understood! I am glad you are watching this. It's easy to mess these things up. Once I have a patch ready (seems it'll have to wait till tomorrow after all *sigh), I hope you will also review it with a critical eye (assuming your time permits, of course).

**Comment by fingolfin** [comment]
This was fixed by 41fc3f21af1395005eaaf95ec7d73946f63b6b06 last August, I just forgot to close this issue.
